### PR TITLE
Various Performance Improvements

### PR DESCRIPTION
**Pull request by pranaygp**
## Summary

This PR includes performance optimizations for the step handler and comprehensive OpenTelemetry tracing improvements.

### Performance Optimizations
1. **Skip initial HTTP call** - Remove redundant `world.steps.get()` call and rely on `step_started` to return the step entity
2. **Parallelize completion operations** - Run `step_completed` event creation and `serializeTraceCarrier()` concurrently
3. **Add server-side retryAfter validation** - Local and postgres worlds now validate retryAfter timestamps and return HTTP 425 when not reached
4. **Lazy ref resolution for fire-and-forget events** - Use `remoteRefBehavior: 'lazy'` for event types where the client discards the response data (e.g. `step_created`, `step_completed`, `step_failed`, `run_completed`), skipping expensive S3 ref resolution (~200-460ms savings per event)

### Bug Fixes
5. **Fix race condition** - Await `step_started` before hydration to ensure correct attempt count in error handlers
6. **Fix HTTP status codes** - Use 409 (Conflict) for terminal state steps instead of 410 across all worlds

### OpenTelemetry Improvements
7. **W3C trace context propagation** - Add `traceparent`/`tracestate` headers to step queue messages for cross-service trace linking
8. **Service attribution** - Add `peer.service` and RPC semantic conventions for Datadog service maps (workflow-server, VQS)
9. **Enhanced span coverage** - Add `step.hydrate`, `step.dehydrate`, `workflow.replay` spans
10. **Improved naming** - Rename `queueMessage` → `queue.publish`, use uppercase WORKFLOW/STEP span names
11. **Baggage propagation** - Propagate workflow context via OTEL baggage (`workflow.run_id`, `workflow.name`)
12. **Milestone events** - Add span events: `retry.scheduled`, `step.skipped`, `step.delayed`
13. **Error telemetry** - Use `recordException()` with error categorization (fatal/retryable/transient)
14. **Event type in span names** - `world.events.create` spans now include event type (e.g., `world.events.create step_started`)
15. **world-local instrumentation** - Add OTEL instrumentation matching world-vercel

**Estimated savings:** 50-80ms per step (one fewer HTTP round-trip) + 200-460ms per fire-and-forget event (skip S3 ref resolution)

## Related
- Requires: https://github.com/vercel/workflow-server/pull/221
- Requires: https://github.com/vercel/workflow-server/pull/231

## Test plan
- [x] Unit tests pass (`pnpm test` - 304 tests passing)
- [x] TypeScript compiles
- [ ] E2E tests with sleep workflow
- [ ] Test edge cases:
  - Step already completed → 409 → re-enqueue workflow
  - retryAfter not reached → 425 → return timeout
  - Max retries exceeded → fail step

## Observability Impact

After these changes, Datadog traces will show:
- **Complete trace linking** across workflow and step executions
- **Service maps** with workflow-server and VQS as external services
- **Detailed timing breakdown** for hydration, execution, and dehydration
- **Error categorization** for easier debugging (fatal vs retryable vs transient)
- **Event type visibility** in span names for filtering

### CODE REVIEW

File: packages/core/src/telemetry.ts [hunk 1]
```diff
@@ -188,3 +188,67 @@ export function linkToCurrentContext(): Promise<[api.Link] | undefined> {
     return [{ context }];
   });
 }
+
+// ============================================================
+// Baggage Propagation Utilities
```
Comment (VaguelySerious) [inline packages/core/src/telemetry.ts:193]: TIL about OTEL baggage. Naming is hard

File: packages/world-vercel/src/events.ts [hunk 2]
```diff
@@ -132,10 +132,23 @@ export async function createWorkflowRunEvent(
   // For run_created events, runId is null - use "null" string in the URL path
   const runIdPath = id === null ? 'null' : id;
 
+  // Determine remoteRefBehavior based on event type:
+  // Events where the client uses the response entity data need 'resolve' (default).
+  // Events where the client discards the response can use 'lazy' to skip expensive
+  // S3 ref resolution on the server, saving ~200-460ms per event.
+  const eventsNeedingResolve = new Set([
```
Comment (VaguelySerious) [inline packages/world-vercel/src/events.ts]: Could be moved outside of the function - maybe event to world package or similar, since the way we except event responses to look like is consistent across worlds. Could be helpful for reference for third-party worlds, mostly, if the comment strings are framed appropriately.

Separately, it looks like we're missing fine-grained control for resolution (e.g. resolve run but not step, resolve step but not run, etc.), which could be a future optimization

File: packages/world-local/src/storage/events-storage.ts [hunk 3]
```diff
@@ -194,7 +194,7 @@ export function createEventsStorage(basedir: string): Storage['events'] {
         if (isStepTerminal(validatedStep.status)) {
           throw new WorkflowAPIError(
             `Cannot modify step in terminal state "${validatedStep.status}"`,
-            { status: 410 }
+            { status: 409 }
```
Comment (VaguelySerious) [inline packages/world-local/src/storage/events-storage.ts:197]: A lot of other similar conflicts in the local world still throw 410. Should those be changed too?

File: packages/core/src/runtime/step-handler.ts [hunk 4]
```diff
@@ -327,26 +366,55 @@ const stepHandler = getWorldHandlers().createQueueHandler(
...
-            // result was dehydrated above by dehydrateStepReturnValue, which returns Uint8Array
-            await world.events.create(workflowRunId, {
-              eventType: 'step_completed',
-              specVersion: SPEC_VERSION_CURRENT,
-              correlationId: stepId,
-              eventData: {
-                result: result as Uint8Array,
-              },
-            });
+            // OPTIMIZATION 2: Parallelize step_completed with trace serialization
```
Comment (VaguelySerious) [inline packages/core/src/runtime/step-handler.ts]: Nit: Some of these comments read more as change indicators rather than current-state docs
```suggestion
```

File: packages/world-vercel/src/events.ts (hunk 2 above)
Comment (pranaygp) [reply packages/world-vercel/src/events.ts]: Moved to module scope — good call, no reason to recreate it on every call.

Moving this to `@workflow/world` so all world implementations can share it, and adding fine-grained per-entity resolution control (e.g. resolve run but not step) are both great follow-up ideas. Will track those separately.

File: packages/core/src/runtime/step-handler.ts (hunk 4 above)
Comment (pranaygp) [reply packages/core/src/runtime/step-handler.ts]: Cleaned up the changelog-style OPTIMIZATION comments in step-handler.ts. They now describe current-state behavior instead of referencing what changed.

File: packages/world-local/src/storage/events-storage.ts (hunk 3 above)
Comment (pranaygp) [reply packages/world-local/src/storage/events-storage.ts:197]: Good question\! I audited all three implementations (workflow-server, world-local, world-postgres) to verify status code consistency.

**Summary of the error semantics:**

| Condition | workflow-server error | HTTP Status | Meaning |
|---|---|---|---|
| Run state transition on terminal run | `InvalidOperationStateError` | **409** Conflict | Can't re-complete/re-fail a finished run |
| Creating step/hook on terminal run | `InvalidOperationStateError` | **409** Conflict | Can't create new work on a finished run |
| Step in terminal state | `InvalidOperationStateError` | **409** Conflict | Step already completed/failed |
| Non-running step on terminal run | Not applicable (falls through) | **410** Gone | Run is done, step wasn't even started |
| Run not in 'running' state | `WorkflowNotRunningError` | **410** Gone | Run hasn't started or is gone |
| retryAfter not reached | `RetryAfterNotReachedError` | **425** Too Early | Wait before retrying |
| Step/run/hook not found | `EntityNotFoundError` | **404** Not Found | Entity doesn't exist |

**Findings:** Both world-local and world-postgres were using 410 (Gone) for cases that should be 409 (Conflict), specifically:
- "Cannot transition run from terminal state" — should be 409 (matches `InvalidOperationStateError` in workflow-server)
- "Cannot create new entities on terminal run" — should be 409 (same reason)  
- step_completed/step_failed fallbacks for step in terminal state (world-postgres only) — should be 409

**The remaining 410s are correct:**
- "Cannot modify non-running step on run in terminal state" = the run is gone, 410 is appropriate (matches `WorkflowNotRunningError` semantics)

All mismatches have been fixed in this PR.
